### PR TITLE
[Fixes #2144] Incorrect target element of nested tag helper

### LINKED ISSUE #2144: Incorrect target element of nested tag helper
The following code:
```csharp
namespace CustomTagHelpers
{
    [HtmlTargetElement("outer-child-tag")]
    [HtmlTargetElement("child-tag", ParentTag = "parent-tag")]
    public class ChildTagHelper : TagHelper
    {
        public override void Process(TagHelperContext context, TagHelperOutput output)
        {
            output.TagName = "span";
            output.Content.SetContent("child content");
        }
    }

    [HtmlTargetElement("parent-tag")]
    [RestrictChildren("child-tag")]
    public class ParentTagHelper : TagHelper
    {
        public override void Process(TagHelperContext context, TagHelperOutput output)
        {
            output.TagName = "div";

            output.Content.SetContent("parent content");
        }
    }
}
```
aims at exposing the `ChildTagHelper` as `<child-tag>` when used inside `<parent-tag>` and as `<outer-child-tag>` on root level.

However, `<child-tag>` is also exposed on root level (screenshot from Visual Studio):
image

Can you provide information why the above is happening and is there a way to update the definition:
```
 [HtmlTargetElement("child-tag", ParentTag = "parent-tag")]
```
so that it restricts the usage of `<child-tag>` in `<parent-tag>` only?

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: test/Microsoft.VisualStudio.Editor.Razor.Test/DefaultTagHelperCompletionServiceTest.cs [hunk 1]
```diff
@@ -838,6 +838,75 @@ public void GetElementCompletions_EnsuresDescriptorsHaveSatisfiedParent()
...
 
+        [Fact]
+        public void GetElementCompletions_NoContainingParentTag_DoesNotGetCompletionForRuleWithParentTag()
+        {
+            // Arrange
+            var documentDescriptors = new[]
+            {
+                TagHelperDescriptorBuilder.Create("Tag1", "TestAssembly")
+                    .TagMatchingRuleDescriptor(rule => rule.RequireTagName("tag-b"))
+                    .TagMatchingRuleDescriptor(rule => rule.RequireTagName("tag-a.1").RequireParentTag("tag-a"))
```
Comment (ajaybhargavb) [inline test/Microsoft.VisualStudio.Editor.Razor.Test/DefaultTagHelperCompletionServiceTest.cs]: Suggest using more descriptive names like `parent`, `child` here. It'll be easier to visualize.

File: src/Microsoft.VisualStudio.Editor.Razor/DefaultTagHelperCompletionService.cs [hunk 2]
```diff
@@ -194,7 +194,7 @@ public override ElementCompletionResult GetElementCompletions(ElementCompletionC
                     {
                         // If there is an unknown HTML schema tag that doesn't exist in the current completion we should add it. This happens for
                         // TagHelpers that target non-schema oriented tags.
-                        addRuleCompletions = true;
+                        addRuleCompletions = TagHelperMatchingConventions.SatisfiesParentTag(completionContext.ContainingTagName, rule);
```
Comment (kichalla) [inline src/Microsoft.VisualStudio.Editor.Razor/DefaultTagHelperCompletionService.cs]: FYI: This is the same condition we initially run to filter the taghelpers https://github.com/aspnet/Razor/blob/release/2.1/src/Microsoft.VisualStudio.Editor.Razor/DefaultTagHelperFactsService.cs#L155

Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Editor.Razor/DefaultTagHelperCompletionService.cs]: The `DefaultTagHelperFactsService` applies this at a broader scale (it doesn't have all of the conditions above). You need to filter every rule from the descriptor by this `SatisfiesParentTag`. Just as an example if you add `[OutputElementHint("span")]` to the `ChildTagHelper` in the originally reported issue it'll ignore the parent restriction again.

**Comment by pranavkm** [comment]
@dotnet-bot help

**Comment by pranavkm** [comment]
@dotnet-bot build Windows Release x64

**Comment by pranavkm** [comment]
@dotnet-bot test Windows Release x64

**Comment by pranavkm** [comment]
@dotnet-bot build Windows Release x64 Build

**Comment by pranavkm** [comment]
@dotnet-bot test Windows Release x64 Build

**Comment by kichalla** [comment]
https://github.com/aspnet/Razor/commit/f831df00a73e33999da6f5df92d08cbc05d12b3f
https://github.com/aspnet/Razor/commit/431a8e195a7261bdf49f63c214b05771c4e4b250

**Pull request merged**